common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -795,13 +796,34 @@ public void registerExecutor(String appId, ExecutorShuffleInfo executorInfo) {
   }
 
   /**
-   * Close the DB during shutdown
+   * Shutdown mergedShuffleCleaner and close the DB during shutdown
    */
   @Override
   public void close() {
+    if (!mergedShuffleCleaner.isShutdown()) {
+      // Use two phases shutdown refer to
+      // https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html
+      try {
+        mergedShuffleCleaner.shutdown();
+        // Wait a while for existing tasks to terminate
+        if (!mergedShuffleCleaner.awaitTermination(10L, TimeUnit.SECONDS)) {
```

Comment:
QQ: Wondering if this timeout should be configurable.
+CC @otterc

Reply:
I also want to know this

Reply:
Also, CC @zhouyejoe

Reply:
https://github.com/apache/spark/blob/06997d6eb73f271aede5b159d86d1db80a73b89f/sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/session/SessionManager.java#L208-L215

In spark, only the above code use a config timeout, the others are hard coded.

Reply:
Yes, the issue here is we are actually waiting for tasks to cleanup on disk files and cleanup the DB - which potentially impacts recovery if we miss out on updates.

Reply:
Got it. Really need @zhouyejoe and @otterc  help for further check, thanks ~

Reply:
IMO this timeout, which is related to shutdown of the service, is not something cluster admin would try to tune. I don't really see much benefits of making it configurable. If the need arises, then it can be made configurable in the future. For now, I would not add another specific config for it.

Reply:
Do you think 10s is sufficient ?

Reply:
The cleanup tasks that this service handles remove the `appAttemptShuffleMergeId` from the db and close and delete all the partition files. The db operation would not take much time. Closing and deleting all the merged shuffle files may take take longer. Maybe we can increase this to a 1 minute?

Reply:
This is why I was wondering if we need to make this configurable :-)
Depending on the cluster load (or rather, number of running applications in cluster which have used this NM as a merger), this time to cleanup + close might be higher or lower - which would depend on cluster utilization characteristics ?

Reply:
I see your point. Though, I still think that such timeouts related to service shutdowns are rarely changed. I am fine either ways though.

Reply:
Let us do that as a follow up then, based on what behavior we see in production.
Can we log the number of `Runnable`'s which did not complete gracefully in that case (returned in `shutdownNow`) ? It will be a proxy for whether this needs to be tuned.

Note that a higher value does not necessarily mean we wait that long - just that we wait upto that in case some of the tasks are taking longer.

Reply:
I'm assuming you mostly care for rolling restart scenario?  I thought they wanted it pretty quick but can't seem to find any config related to timeouts there.

Reply:
this is not a daemon thread then?  I guess you really want this to complete before shutdown it seems?  I would definitely make it configurable then even if its a not documented config

Reply:
[34605f9](https://github.com/apache/spark/pull/37624/commits/34605f9b37e132557fd70d70d2fc2393b69ecab3) do the following:

1. make the timeout as a new config `spark.shuffle.push.server.mergedShuffleCleaner.shutdown.timeout`but not  documented in this pr
2. bump the timeout default value up to 60s.

Reply:
Agree upon the point to make it configurable, depending on the shuffle load in the cluster.

Reply:
QQ: For the second mergedShuffleCleaner.awaitTermination(), do we also need to wait 60 seconds? Or we can hardcode a shorter time period there?

Reply:
Let us keep it consistent @zhouyejoe - particularly given we are not really expecting the second to take that long; unless it is blocked in something external (like db write/delete)

Reply:
Make sense